File: easy/python/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for x in s:
            if x == "{":
                stack.append("}")
            elif x == "[":
                stack.append("]")
            elif x == "(":
                stack.append(")")
            elif stack and x == stack.pop():
                continue
            else:
                return False

        return len(stack) == 0

File: easy/ts/valid_parentheses.ts
function isValid(s: string): boolean {
  const stack: string[] = [];
  const mapping: { [key: string]: string } = {
    '(': ')',
    '{': '}',
    '[': ']',
  }
  for (let i = 0; i < s.length; i++) {
    const char = s[i];
    if (mapping[char]) {
      stack.push(mapping[char]);
    } else if (stack.pop() === char) {
      continue;
    } else {
      return false;
    }
  }

  return stack.length == 0
}

function isValid2(s: string): boolean {
  const closingStack: string[] = [];
  const openingBrackets = {
    "(": ")",
    "{": "}",
    "[": "]",
  }
  for (const str of s) {
    if (openingBrackets[str]) {
      closingStack.push(openingBrackets[str]);
    } else {
      const expectedClosing = closingStack.pop();
      if (expectedClosing !== str) return false;
    }
  }

  return closingStack.length === 0;
}